### DISCUSSION TITLE
When power restored relay "blink" with poweronstate 1

### DISCUSSION THREAD

**Discussion by Baton34** [arendst/Tasmota#13519] [Q&A] [answered]
I have smtonoff wifi switch with energy meter. I use 'sonoff pow r2' template and everything works fine. But when power is removed and then restored, the relay remains on but after some time (about 1 second) relay switch off and then immediately on.
So controlled device powered on, off and on again.
How to stop "blinking" device on power restored?
tasmota 10.0.0
powerOnstate 1
SaveData 1
SetOption63 On and Off did not change anything

**Comment by barbudor** [comment]
Setting Weblog 3 and looking at the logs may show some hints on what his happenning

**Reply by sfromis** [reply]
One source of "unexpected actions" is using retained messages when they should not have been

**Comment by Baton34** [comment]
Here is weblog on restoring power:
`00:00:00.001 HDW: ESP8285
00:00:00.050 CFG: Загружено из флэш-памяти F5, Подсчет 190
00:00:00.055 SER: Set to 8E1 4800 bit/s
00:00:00.059 QPC: Count 1
00:00:00.061 CFG: CR 357/699, Busy 0
00:00:00.065 ROT: Mode 1
00:00:00.066 SRC: Restart
00:00:00.067 Проект tasmota - Tasmota Версия 10.0.0(tasmota)-2_7_4_9(2021-10-19T08:26:03)
00:00:00.071 SER: Set to 8E1 4800 bit/s
00:00:00.074 SNS: Hardware Serial
00:00:00.227 WIF: Попытка подключения...
00:00:00.563 WIF: Подключение к AP1 bWIFI Channel 9 BSSId  в режиме 11n как tasmota-E7BAF1-6897...
00:00:01.527 WIF: Попытка подключения...
00:00:02.502 WIF: Попытка подключения...
00:00:03.517 WIF: Соединен
00:00:03.756 HTP: Веб-сервер активен tasmota-E7BAF1-6897 с IP-адресом 172.30.1.19
00:00:03.759 RSL: INFO1 = {"Info1":{"Module":"Sonoff Pow R2","Version":"10.0.0(tasmota)","FallbackTopic":"cmnd/DVES_E7BAF1_fb/","GroupTopic":"cmnd/tasmotas/"}}
00:00:03.762 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-E7BAF1-6897","IPAddress":"172.30.1.19"}}
00:00:03.764 RSL: INFO3 = {"Info3":{"RestartReason":"Power On"}}
00:00:03.766 RSL: RESULT = {"POWER":"ON"}
00:00:03.767 RSL: POWER = ON
00:00:04.442 RTC: UTC 2021-10-30T13:28:49, DST 2021-03-28T02:00:00, STD 2021-10-31T03:00:00
16:28:51.451 QPC: Reset
16:28:53.367 APP: Количество загрузок 39
16:28:53.372 RSL: STATE = {"Time":"2021-10-30T16:28:53","Uptime":"0T00:00:10","UptimeSec":10,"Heap":28,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"ON","Wifi":{"AP":1,"SSId":"WIFI","BSSId":"","Channel":9,"Mode":"11n","RSSI":92,"Signal":-54,"LinkCount":1,"Downtime":"0T00:00:05"}}
16:28:53.380 RSL: SENSOR = {"Time":"2021-10-30T16:28:53","ENERGY":{"TotalStartTime":"2021-10-28T00:00:00","Total":0.002,"Yesterday":0.000,"Today":0.002,"Period": 2,"Power":95,"ApparentPower":95,"ReactivePower": 0,"Factor":1.00,"Voltage":234,"Current":0.407}}
16:28:54.386 CFG: Сохранено во флэш-память F4, Подсчет 191, Байт 4096`

**Reply by barbudor** [reply]
Can't see a power off here.
If it ain't software. It's hardware and beyond Tasmota control. 

And you didn't set Weblog 3

**Reply by Baton34** [reply]
> beyond Tasmota control.
Maybe this is the case. 

> And you didn't set Weblog 3
I'm pretty sure I did it. 
`00:00:00.001 HDW: ESP8285
00:00:00.050 CFG: Загружено из флэш-памяти F9, Подсчет 194
00:00:00.055 SER: Set to 8E1 4800 bit/s
00:00:00.059 QPC: Count 1
00:00:00.061 CFG: CR 357/699, Busy 0
00:00:00.065 ROT: Mode 1
00:00:00.066 SRC: Restart
00:00:00.068 Проект tasmota - Tasmota Версия 10.0.0(tasmota)-2_7_4_9(2021-10-19T08:26:03)
00:00:00.072 SER: Set to 8E1 4800 bit/s
00:00:00.075 SNS: Hardware Serial
00:00:00.227 WIF: Попытка подключения...
00:00:00.563 WIF: Подключение к AP1 WIFI Channel 9 BSSId  в режиме 11n как tasmota-E7BAF1-6897...
00:00:01.527 WIF: Попытка подключения...
00:00:02.504 WIF: Попытка подключения...
00:00:03.519 WIF: Соединен
00:00:03.755 HTP: Веб-сервер активен tasmota-E7BAF1-6897 с IP-адресом 172.30.1.19
00:00:03.759 RSL: INFO1 = {"Info1":{"Module":"Sonoff Pow R2","Version":"10.0.0(tasmota)","FallbackTopic":"cmnd/DVES_E7BAF1_fb/","GroupTopic":"cmnd/tasmotas/"}}
00:00:03.761 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-E7BAF1-6897","IPAddress":"172.30.1.19"}}
00:00:03.763 RSL: INFO3 = {"Info3":{"RestartReason":"Power On"}}
00:00:03.765 RSL: RESULT = {"POWER":"ON"}
00:00:03.766 RSL: POWER = ON
00:00:04.443 RTC: UTC 2021-10-30T13:46:29, DST 2021-03-28T02:00:00, STD 2021-10-31T03:00:00
16:46:31.451 QPC: Reset
16:46:33.366 APP: Количество загрузок 41
16:46:33.371 RSL: STATE = {"Time":"2021-10-30T16:46:33","Uptime":"0T00:00:10","UptimeSec":10,"Heap":28,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"ON","Wifi":{"AP":1,"SSId":"WIFI","BSSId":"","Channel":9,"Mode":"11n","RSSI":88,"Signal":-56,"LinkCount":1,"Downtime":"0T00:00:05"}}
16:46:33.379 RSL: SENSOR = {"Time":"2021-10-30T16:46:33","ENERGY":{"TotalStartTime":"2021-10-28T00:00:00","Total":0.002,"Yesterday":0.000,"Today":0.002,"Period": 2,"Power":92,"ApparentPower":92,"ReactivePower": 0,"Factor":1.00,"Voltage":229,"Current":0.403}}
16:46:34.393 CFG: Сохранено во флэш-память F8, Подсчет 195, Байт 4096`
What I have to see to understand that this is a weblog 3 ?

**Reply by barbudor** [reply]
Just type Weblog without parameter to see which level you are on
Weblog 3 would have much more logs including setting/flash informations

**Reply by Baton34** [reply]
> Just type Weblog without parameter to see which level you are on

Nothing changed:
`00:00:00.001 HDW: ESP8285
00:00:00.050 CFG: Загружено из флэш-памяти F8, Подсчет 195
00:00:00.055 SER: Set to 8E1 4800 bit/s
00:00:00.059 QPC: Count 2
00:00:00.061 CFG: CR 357/699, Busy 0
00:00:00.065 ROT: Mode 1
00:00:00.066 SRC: Restart
00:00:00.068 Проект tasmota - Tasmota Версия 10.0.0(tasmota)-2_7_4_9(2021-10-19T08:26:03)
00:00:00.071 SER: Set to 8E1 4800 bit/s
00:00:00.074 SNS: Hardware Serial
00:00:00.226 WIF: Попытка подключения...
00:00:00.563 WIF: Подключение к AP1 WIFI Channel 9 BSSId  в режиме 11n как tasmota-E7BAF1-6897...
00:00:01.528 WIF: Попытка подключения...
00:00:02.504 WIF: Попытка подключения...
00:00:03.518 WIF: Соединен
00:00:03.754 HTP: Веб-сервер активен tasmota-E7BAF1-6897 с IP-адресом 172.30.1.19
00:00:03.758 RSL: INFO1 = {"Info1":{"Module":"Sonoff Pow R2","Version":"10.0.0(tasmota)","FallbackTopic":"cmnd/DVES_E7BAF1_fb/","GroupTopic":"cmnd/tasmotas/"}}
00:00:03.760 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-E7BAF1-6897","IPAddress":"172.30.1.19"}}
00:00:03.762 RSL: INFO3 = {"Info3":{"RestartReason":"Power On"}}
00:00:03.764 RSL: RESULT = {"POWER":"ON"}
00:00:03.766 RSL: POWER = ON
00:00:04.507 RTC: UTC 2021-10-30T19:56:46, DST 2021-03-28T02:00:00, STD 2021-10-31T03:00:00
22:56:48.452 QPC: Reset
22:56:50.367 APP: Количество загрузок 42
22:56:50.372 RSL: STATE = {"Time":"2021-10-30T22:56:50","Uptime":"0T00:00:10","UptimeSec":10,"Heap":29,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"ON","Wifi":{"AP":1,"SSId":"WIFI","BSSId":"","Channel":9,"Mode":"11n","RSSI":92,"Signal":-54,"LinkCount":1,"Downtime":"0T00:00:05"}}
22:56:50.380 RSL: SENSOR = {"Time":"2021-10-30T22:56:50","ENERGY":{"TotalStartTime":"2021-10-28T00:00:00","Total":0.002,"Yesterday":0.000,"Today":0.002,"Period": 2,"Power":95,"ApparentPower":95,"ReactivePower": 0,"Factor":1.00,"Voltage":234,"Current":0.405}}
22:56:51.384 CFG: Сохранено во флэш-память F7, Подсчет 196, Байт 4096
22:57:10.801 HTP: Главное Меню
22:57:12.351 HTP: Консоль
22:57:16.959 CMD: weblog
22:57:16.961 SRC: WebConsole from 172.30.0.2
22:57:16.963 CMD: Grp 0, Cmnd 'WEBLOG', Idx 1, Len 0, Data ''
22:57:16.966 RSL: RESULT = {"WebLog":3}`

**Reply by barbudor** [reply]
Ok, sorry but you copied the logs without return to line, so it is hard to read. You are indeed in Weblog 3

Still I can't see any POWER OFF event in the logs, so the blinking is not due from software.
I don't know the schematic of the POW R2, but if it is similar to the original POW:
![image](https://user-images.githubusercontent.com/2996042/139558018-5dbc7b8e-4f2b-4c86-a85b-16bf99f0a5ee.png)

It should be possible to fix your problem by removing the R22 pull-down at the gate of the MOSFET Q4 and move it (or replace) as a pull-up to 3V3 (not to 5V !!) so that if the  GPIO becomes floating at some point during the boot (before Tasmota set it HIGH) the pull-up would maintain the Relay ON.

**Reply by Baton34** [reply]
I got your idea, but I got a related question: 
Is it possible that tasmota, at boot time, pull-down gpio12 (to which the relay is connected) and then, according to "poweronstate 1", feeds 3.3V to gpio12, and therefore an on-off-on sequence is obtained? Maybe I need to change initial state from low to high of the gpio12 in tasmota initializing sequence in it source code?
Thanks for the help

**Reply by sfromis** [reply]
Testing a Sonoff Pow R2 with `PowerOnState 1` on Tasmota v10.0.0, and removing/restoring input power, I got the expected behavior of it turning on with no extra "blink". Had no need for hanky-panky with resistors or code change.

Note that removing input power is not a good combination with an energy monitoring device, as the consumption data since last save will be lost.

**Reply by Baton34** [reply]
I have a different device - Smtonoff which works with Sonoff Pow R2 template and has a different schema. Its relay is driven by a BL8023H bi-directional relay driver which is driven by an undefined IC driven by gpio12 from esp8285. Thats why I asked about initial condition of gpio12 in tasmota source. I want to try software fix before trying hardware fix.

**Reply by sfromis** [reply]
Since I have no blink, with the relay directly controlled by gpio12, this suggests that the issue is not with the Tasmota control of that pin, and I see no clear "opportunities for fixing" in software.

At restart/boot time (without powering off), Tasmota is able to maintain a stable signal on gpio12 without any blinking, even during updates from one Tasmota version to the next. My main desktop computer is behind a Sonoff Pow R2, and years ago it could crash during such an upgrade cycle, till this was improved.

**Comment by barbudor** [comment]
So your likely having 1 or 2 sources of unknown hardware behavior between Tasmota on the ESP and your relay.
There's nothing software can do for you.

**Comment by Baton34** [comment]
Ok. I understood, thank you.

**Comment by Baton34** [comment, marked as answer]
So I fixed my problem.
As far as I understand, this unidentified controller that controls the relay, upon initialization, puts a relay disconnecting signal at its output.
Therefore, when the external power is applied (when it was turned off, the relay was in the on state), the load is turned on. Then the unknown controller is initialized and gives a signal to turn off the relay. Then esp is loaded and gives a signal to turn on the relay - we get a blink.
What I've done? I connected the relay with reverse polarity and changed the tasmota gpio12 template to relay_i. Now, when external power is applied, the controller gives a signal to turn off, but the relay is connected with reverse polarity and remains on, then esp is loaded and gives a reverse signal to turn on the relay, but nothing changes for the controller and the relay remains in the same position - on. That's it, no blinking.

Here are the photos of the device:
![0](https://user-images.githubusercontent.com/10797316/139685130-11fdce15-319a-4a4a-8552-d449171819e3.jpg)
![1](https://user-images.githubusercontent.com/10797316/139685141-2d33e989-97d9-4e2d-9cdf-5a114555a47e.jpg)
![2](https://user-images.githubusercontent.com/10797316/139685172-1460ccd7-ed23-407c-aafd-b07224ac3735.jpg)

**Reply by sfromis** [reply]
Nice solution to get around the quirks with that device.
